### PR TITLE
Do not emit invalid suggestion in E0191 when spans overlap

### LINKED ISSUE #115019: ICE: debug assertions: E0191 overlapping spans
### Code

```Rust
pub struct Flatten<I> {
    inner: <IntoIterator<Item: IntoIterator<Item: >>::IntoIterator as Item>::core,
}
```

### Meta

`rustc --version --verbose`:
```
6ef7d16be0fb9d6ecf300c27990f4bff49d22d46
```

### Error output

```
error[E0191]: the value of the associated types `IntoIter` (from trait `IntoIterator`), `IntoIter` (from trait `IntoIterator`), `Item` (from trait `IntoIterator`), `Item` (from trait `IntoIterator`) must be specified
 --> 58279A35D0EC648B4ACA3DAFB71E174BF71401B6729EDFD500DDEF2950F94C0C.rs:2:13
  |
2 |     inner: <IntoIterator<Item: IntoIterator<Item: >>::IntoIterator as Item>::core,
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |             |                  |
  |             |                  associated types `Item`, `IntoIter` must be specified
  |             associated types `Item`, `IntoIter` must be specified
  |
help: specify the associated types
  |
2 |     inner: <IntoIterator<Item: IntoIterator<Item: >, Item = Type, IntoIter = Type>IntoIterator<Item: , Item = Type, IntoIter = Type>>::IntoIterator as Item>::core,
  |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
span

</p>
</details>

### PR DESCRIPTION
**Pull request by estebank**
Fix #115019.

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/astconv/errors.rs [hunk 1]
```diff
@@ -597,7 +597,28 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 }
             }
         }
-        if !suggestions.is_empty() {
+        suggestions.sort_by_key(|&(span, _)| span);
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/errors.rs:600]: Are you certain that the suggestion spans will end up sorted in such a way that two non-adjacent keys overlap? If not, then since this is the error path, I feel like it may just be easier to do the O(N^2) full pairwise comparison here.

File: compiler/rustc_hir_analysis/src/astconv/errors.rs [hunk 2]
```diff
@@ -597,7 +597,28 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        //             associated types `Item`, `IntoIter` must be specified
+        // }
+        let mut overlaps = false;
+        let mut sugg = suggestions.iter().peekable();
+        while let Some((span, _)) = sugg.next() {
+            let Some((next, _)) = sugg.peek() else {
+                break;
+            };
+            overlaps |= span.overlaps(*next);
+        }
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/errors.rs:620]: I think this loop can be simplified? Will need to be tweaked if the above assumption is not true, though.

```suggestion
        let overlaps = suggestions.windows(2).any(|pair| pair[0].0.overlaps(pair[1].0));
```

**Comment by compiler-errors** [comment]
Please squash this ideally if you apply any of my suggestions.

**Comment by compiler-errors** [comment]
r=me after confirming whether the assumption about overlapping spans is true or false, then either applying the simplification suggested above or just changing it to an O(N^2) overlap check like:

```
let overlaps = suggestions.iter().any(|(span, _)| suggestions.iter().any(|(other_span, _)| span != other_span && span.overlaps(other_span)));
```

or something like that.

File: compiler/rustc_hir_analysis/src/astconv/errors.rs (hunk 1 above)
Comment (estebank) [reply compiler/rustc_hir_analysis/src/astconv/errors.rs:600]: `SpanData` manually implements `Ord` on `(lo, hi, ctxt)`. That means that by sorting any two adjacent spans that don't overlap imply that the spans further apart from each other also don't overlap.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/astconv/errors.rs:600]: What about spans that overlap only at their prefix? Like spans that look like: `[(1, 5), (2, 3), (4, 5)]`, where the first and last elements overlap?

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/astconv/errors.rs:600]: Maybe it doesn't matter, since by construction that would imply any intermediate spans overlap. Idk.

Comment (estebank) [reply compiler/rustc_hir_analysis/src/astconv/errors.rs:600]: Your last comment is exactly right.

File: compiler/rustc_passes/src/reachable.rs [hunk 3]
```diff
@@ -101,7 +101,7 @@ impl<'tcx> Visitor<'tcx> for ReachableContext<'tcx> {
                     // If this path leads to a constant, then we need to
                     // recurse into the constant to continue finding
                     // items that are reachable.
-                    Res::Def(DefKind::Const | DefKind::AssocConst, _) => {
+                    Res::Def(DefKind::Const | DefKind::AssocConst | DefKind::Static(_), _) => {
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/reachable.rs:104]: Please revert this change. I assume you're planning on putting up another PR for that reachability bug.

Comment (estebank) [reply compiler/rustc_passes/src/reachable.rs:104]: Damn, I must have accidentally `git commit -a` at some point 🤦
